File: 二叉树/654最大二叉树.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode constructMaximumBinaryTree(int[] nums) {
        return traversal(nums, 0, nums.length-1);
    }
    public TreeNode traversal(int[] nums, int L, int R) {
        //终止条件
        if(L > R) {
            return null;
        }
        int maxIndex = L;
        for(int i = L; i <= R; i++) {
            if(nums[i]>nums[maxIndex]){
                maxIndex = i;
            }
        }
        TreeNode root = new TreeNode(nums[maxIndex]); //中
        root.left = traversal(nums, L, maxIndex-1);   // 左
        root.right = traversal(nums, maxIndex+1, R); // 右
        return root;
    }
}

File: 二叉树/654. 最大二叉树.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def constructMaximumBinaryTree(self, nums: List[int]) -> TreeNode:
        def constructTree(nums, L, R):
            if L > R: return None
            maxNum = L
            for i in range(L+1, R+1):
                if nums[i] > nums[maxNum]:
                    maxNum = i
                    
            root = TreeNode(nums[maxNum])
            root.left = constructTree(nums, L, maxNum-1)
            root.right = constructTree(nums, maxNum+1, R)
            return root

        return constructTree(nums, 0, len(nums)-1)
